### ISSUE TITLE
Table rendering and ANRs

### ISSUE THREAD

**Issue by k0shk0sh** [noties/Markwon#150]
* **Markwon version**: _{4.0.2}_

I had plans since a year integrating this lib in FastHub, matter of fact I have actually forked it back in 2017 https://github.com/k0shk0sh/FastHub/blob/ee2c95dec4873633abdfc26eb0895929915edd90/app/src/main/java/com/fastaccess/github/platform/markwon/mention/MentionExtension.kt#L12 trying to integrate `@`, `#`, `:emoji:` because I notice that's the only thing missing and its still, but with the possibility of adding own plugin, that became possible and finally decided its time to switch to this lib, you done amazing work and can't even think of how troublesome it is to implement this lib :D. Enough complements :p, I got an issue when using `TablePlugin` in a large issue description where it contains multiple tables specifically this: https://github.com/k0shk0sh/FastHub/issues/823 . is there a better way of loading the tables in a better way to avoid ANRs?

**Comment by noties** [comment]
Hello @k0shk0sh !

Thank you for kind words 😄 !

I could not obtain raw markdown text from the issue that you'd linked (even if you do `quote reply` tables are removed from the markdown). Anyway. There is a lot of text. And by default `Markwon` uses single TextView to display everything. I would suggest trying the `markwon-recycler` artifact, that uses RecyclerView and a TextView for each markdown block. This also comes with few benefits like custom widgets for certain blocks. For example `markwon-recycler-table` artifact uses Android `TableLayout` to render markdown tables (which also allows any markdown to be present inside a table, as default implementation, for example, does not support images inside tables). Also, you could make code blocks scrollable. There is a [sample] that shows how to use it. You can actually start with a simple `markwon-recycler` artifact and see if ANR still persist.

As for `@`, `#` and emojis CorePlugin (since `4.0.0`) has `OnTextAddedListener` which is perfect for these. It skips commonmark tree node manipulation and works directly with text (code and code-blocks are ignored). For example, your plugin can do something like this (taken from `markwon-linkify`):

```java
@Override
public void configure(@NonNull Registry registry) {
    registry.require(CorePlugin.class, new Action<CorePlugin>() {
        @Override
        public void apply(@NonNull CorePlugin corePlugin) {
            corePlugin.addOnTextAddedListener(new LinkifyTextAddedListener(mask));
        }
    });
}
```

There is a small dedicated documentation [section] covering this.

[sample]: https://github.com/noties/Markwon/blob/master/sample/src/main/java/io/noties/markwon/sample/recycler/RecyclerActivity.java
[section]: https://noties.io/Markwon/docs/v4/core/core-plugin.html#ontextaddedlistener

**Comment by k0shk0sh** [comment]
Hi @noties thanks for your response I have integrated `recycler` module and it was very easy! awesome!!! https://github.com/k0shk0sh/FastHub/pull/2599/commits/aae3d0fd09f6ccc037313eec51e13e2ecc1d241b

regarding the ANR its way faster now, that's awesome, but still could see a bit of a lag, its understandable tho and i'll work on this once I finalize my branch. 

i reckon I have to share one recyclerPool because now each `commentROW` contains a recyclerview which is embedded in a parent recyclerview. 

Thanks again

**Comment by k0shk0sh** [comment]
It came to attention that selecting text now becomes hard because each text is rendered separately. Any clue on how to solve this?

**Comment by noties** [comment]
Hello @k0shk0sh !

Hm, if you are using a recycler-view for each comment - yeah a pool is a must. But I find it a bit too much. Maybe you can introduce multiple comment types, like a big one (based on length or something) and a small one, where big one will use recycler-view and a small one will use just a single text-view. Maybe.

Also, you should know that default entry in `markwon-recycler` parses markdown on first on-bind (then caches it). It can cause some trouble. Also I see that you are using `markwon-linkify` plugin, it's really slow, maybe you can try without it first and see if it's the one that slows down the rendering?

Now, text selection... I have to warn you, it's not going to be easy. I have found a really old project that showcases the usage of text selection between multiple text-views in a recycler-view. You can look at it [here](https://github.com/qw1nz/TextSelection/tree/master/app/src/main/java/com/dwdm/textselection). There is also an [article](https://habr.com/ru/company/true_engineering/blog/268501/) about this implementation with the walk-through, but it's in Russian, so I do not know if it will be of any help to you. In short - you will have to take the control of selection in your hands:
* calculate text selection start (which word, position on screen)
* draw text-selection handles
* control the dragging of text-selection handles
* highlight text-selection

Please note that you also won't be able to display system text selection popup... You will have to _mimic_ it somehow.

Maybe your comments/posts can have a dedicated `copy` button instead? This will also allow you to have original markdown copied instead of _rendered_ one. As there is a big chance that original and rendered texts will differ.

**Comment by noties** [comment]
I did manage to obtain source markdown for the issue that you'd shared via Github API. ANR can appear when actual `textView.setText` is called. This can be mitigated if [PrecomputedText] (or [PrecomputedTextCompat]) is used. Currently `Markwon` does not allow to use it (sets text by itself), but I think I will add a way to do it. 

BTW, quick test shows that `PrecomputedText` does not work with `TableLayout` in a `RecyclerView` 🤦‍♂ . And `PrecomputedTextCompat` works only for API >= 21, so on lower devices ANR will still occur.

Anyway, even with this feature added I still think that `RecyclerView` is a way to go when dealing with lengthy markdown content.

[PrecomputedText]: https://developer.android.com/reference/android/text/PrecomputedText
[PrecomputedTextCompat]: https://developer.android.com/reference/androidx/core/text/PrecomputedTextCompat

**Comment by k0shk0sh** [comment]
doing the copy functionality myself is overkill if you ask me :D. 
i'm with going with `PrecomputedTextCompat` to simply avoid having multiple recyclerviews in a recyclerview especially when we have to call notify changes in every viewholder. 

can't wait to see how you integrated it.

Thanks alot for your dedication on this :)

**Comment by noties** [comment]
Hello @k0shk0sh !

I have published the `4.1.0-SNAPSHOT` ([instructions](https://noties.io/Markwon/docs/v4/install.html#snapshot)) version with the `PrecomputedTextSetterCompat` (it will require you to have `androidx.core:core` in dependencies. I also have updated [samples](https://github.com/noties/Markwon/blob/develop/sample/src/main/java/io/noties/markwon/sample/precomputed/PrecomputedActivity.java) to include this functionality. In short:

```java
final Markwon markwon = Markwon.builder(this)
         // your plugins
        .textSetter(PrecomputedTextSetterCompat.create(Executors.newCachedThreadPool()))
        .build();
```

Please give it a try to see if it fits your needs.

I also wanted to mention, that copying text from a table (if used in a TextView not a RecyclerView), won't work (nothing will be copied). Also these tables do not support images.

**Comment by k0shk0sh** [comment]
@noties that was quick!! I'll be home in two hours or so! I'll try it out and update you.

Thanks for taking care of this <3

**Comment by noties** [comment]
You are welcome! 🙌

**Comment by k0shk0sh** [comment]
@noties unforetunely that didn't work for the issue I linked previously, it lagged alot. 
I removed the linkify plugin as well but that didn't work. 

to give you more insight: 

```
 @PerFragment @Provides fun provideMarkwon(
        context: Context,
        preference: FastHubSharedPreference
    ): Markwon = Markwon.builder(context)
        .usePlugin(JLatexMathPlugin.create(context.resources.getDimension(R.dimen.abc_text_size_subhead_material)))
        .usePlugin(TaskListPlugin.create(context))
        .usePlugin(HtmlPlugin.create())
        .usePlugin(GlideImagesPlugin.create(context))
        .usePlugin(TablePlugin.create(context))
        .usePlugin(StrikethroughPlugin.create())
        .usePlugin(
            SyntaxHighlightPlugin.create(
                Prism4j(GrammarLocatorDef()), if (ThemeEngine.isLightTheme(preference.theme)) {
                    Prism4jThemeDefault.create()
                } else {
                    Prism4jThemeDarkula.create()
                }
            )
        )
        .textSetter(PrecomputedTextSetterCompat.create(Executors.newCachedThreadPool()))
        .build()
```
and using it in https://github.com/k0shk0sh/FastHub/blob/4590a4c426effbe385294ccdd6b81c2284699b04/app/src/main/java/com/fastaccess/github/ui/modules/issue/fragment/IssueFragment.kt#L261

I still prefer the textview over  the recyclerview implementation tho 🤷‍♂.

**Comment by noties** [comment]
Hey @k0shk0sh !

The lag can also be caused by the parsing stage. You actually can parse markdown in background, then apply it in the main thread:

```java
// in background
final Spanned spanned = markwon.toMarkdown(md);

// in main thread
markwon.setParsedMarkdown(textView, spanned);
```

**Comment by k0shk0sh** [comment]
so I tried doing that but setParsedMarkdown was still given me the same lag. 
i'm going with this current implementation for now to further my development but I'll get back to it before releasing and try to figure out what's the best way!

```
class MarkwonLoaderLiveData(
    private val markwon: Markwon,
    private val md: String,
    private val scheduler: SchedulerProvider
) : LiveData<Spanned>() {

    private var disposable: Disposable? = null

    override fun onActive() {
        super.onActive()
        disposable = Observable.fromPublisher { s: Subscriber<in Spanned> ->
            runCatching {
                markwon.toMarkdown(md)
            }
                .onFailure { s.onError(it) }
                .onSuccess { s.onNext(it) }
            s.onComplete()
        }
            .subscribeOn(scheduler.ioThread())
            .observeOn(scheduler.uiThread())
            .subscribe({ postValue(it) }, {
                postValue(SpannableString(md)) // return original in case of failure
                it.printStackTrace()
            })
    }

    override fun onInactive() {
        super.onInactive()
        val disposable = disposable
        if (disposable?.isDisposed == false) {
            disposable.dispose()
        }
    }
}
```

**Comment by noties** [comment]
Hello @k0shk0sh !

I did small investigation. Here is what I did:

* Manually set `MovementMethod`
* Manually set `NoCopySpannableFactory`
* change `breakStrategy` and `hyphenationFrequency`
* update library to defer text-view invalidation 😉 

```java
// call these when your view is inflated/created (should be done
// only once and before applying markdown
textView.setMovementMethod(LinkMovementMethod.getInstance());
textView.setSpannableFactory(NoCopySpannableFactory.getInstance());
```

```xml
<!-- rest attributes are omitted -->
<TextView
    android:breakStrategy="simple"
    android:hyphenationFrequency="none" />
```

and then I updated how text-view is invalidated. I defer the invalidation via `textView.post` which leads to a single invalidation call instead of multiple ones. I tested this and performance is much better even without `PrecomputedTextSetter`. You can check latest `4.1.0-SNAPSHOT` to see if it solves the lag problem

**Comment by k0shk0sh** [comment]
@noties sure I'll try this next week when im back from vacation.

Thanks again for the continues help

**Comment by noties** [comment]
A small update: there is no need to use snapshot version as I had released the `4.1.0` version

**Comment by k0shk0sh** [comment]
that really did the trick!, its bloody fast right now, well done really!! 

I think this issue can be closed now :)

**Comment by k0shk0sh** [comment]
since this issue turned into discussion :D I have another question:  I just noticed that gifs doesn't play automatically using glide module. 

Do I need to create/override the plugin to play gifs or there is something i'm missing? as far as I recall, reading the wiki only mentioned gifs if we are using the lib image loading instead of glide/picasso

**Comment by noties** [comment]
Hey @k0shk0sh !

I'm really glad that it did the trick!

About the images: let's create a new issue. Please specify what config do you use. Do the GIF play at all with Glide?
